docs/bricks_design.rst

```diff
@@ -0,0 +1,602 @@
+    >>>         for dim, activation, weight_matrix, bias_vector in zip(
+    >>>                 self.dims[1:], self.activations, self.weight_matrices, self.bias_vectors):
+    >>>             outputs = activation(outputs.dot(weight_matrix) + bias_vector)
+    >>>         return outputs    
+    >>> 
+    >>>     
+    >>> class RecurrentNetwork(Stone):
+    >>>     def __init__(self, dim, weight_initializer):
+    >>>         self.dim = dim
+    >>>         self.weight_initializer = weight_initializer
+    >>>         self.weight_matrix = theano.shared(weight_initializer((dim, dim)))
+    >>>         
+    >>>     def apply_one_step(self, states, inputs, weight_matrix): 
+    >>>         return states.dot(weight_matrix) + inputs
+    >>> 
+    >>>     def apply(self, inputs):
+    >>>         outputs, _ = theano.scan(
+    >>>             self.apply_one_step,
+    >>>             sequences=[inputs], non_sequences=[self.weight_matrix],
+    >>>             outputs_info=[tensor.zeros_like(inputs[0])])
+    >>>         return outputs
+    >>> 
+    >>> 
+    >>> inputs = tensor.matrix('x')
+    >>> inputs.tag.test_value = numpy.random.uniform(size=(10, 754))
+    >>> dims = [754, 100, 100, 10]
+    >>> activations = [tensor.tanh, tensor.tanh, tensor.nnet.softmax]
+    >>> weight_initializers = [
+    >>>     lambda shape: numpy.random.uniform(size=shape)] * len(activations)
+    >>> bias_initializers = weight_initializers
+    >>> stone = MultiLayer(
+    >>>     dims, activations, weight_initializers, bias_initializers)
+    >>> outputs_multilayer = stone.apply(inputs)
+    >>>     
+    >>> inputs = tensor.tensor3('x')
+    >>> inputs.tag.test_value = numpy.random.uniform(size=(5, 10, 100))
+    >>> stone = RecurrentNetwork(100, lambda shape: numpy.random.uniform(size=shape))
+    >>> outputs_recurrent = stone.apply(inputs)
+
+We do not list the hypothetical sampling code here for brevity, but
+hopefully people that have tried to implement such things see that it
+can be implemented using the ``apply_one_step`` method of the ``stone``
+object. It is not such a big deal to save the user from reimplementing
+``apply_one_step`` for a simple recurrent network, but this could just
+as well be LSTM or GRU or even Clockwork RNN, for which
+``apply_one_step`` would be much less trivial.
+
+Note, how that we deliberately avoid the name _layer_ for our
+components. A layer is an element of a sequence, and many of interesting
+modern DL models can hardly be represented as sequences of components
+(consider Neural Turing Machines, Memory Networks, attention-equipped
+Encoder-Decoders with and others).
+
+Similarly, we avoid the wide-spread concept of **the** input of a layer.
+Many components of the systems mentioned in the previous paragraph take
+many inputs, e.g. an attention mechanism often uses both the state of
+the decoder and the encoder input.
+
+Step by step, we have arrived to something virtually indistinguishable
```

Comment:
This is not needed in a tutorial introduction - lets assume people don't know about other projects, unless they are very popular (like Caffe, or PL2)